### ISSUE TITLE
`WidgetsBinding.drawFrame` falsely assumes that `sendFramesToEngine` won't turn from true to false

### ISSUE THREAD

**Issue by dkwingsmt** [flutter/flutter#144261]
During `WidgetsBinding.drawFrame`, a `firstFrameCallback` is scheduled to be called on the next timing report. The `firstFrameCallback` asserts that `sendFramesToEngine` is true. While this assertion might seem reasonable at a glance, since a timing report is called only after `Rasterizer` renders a frame, which requires `sendFramesToEngine` to be true, it's wrong. The assertion can be triggered if `sendFramesToEngine` turns from true to false between when the frame is sent to the engine and when the timing report callback is invoked.

The case where `sendFramesToEngine` turns false right before `firstFrameCallback` is fired is extremely rare. However, the assumption is wrong anyway. We can assert that the `sendFramesToEngine` value when the callback is registered must be true. We must not assert that it hasn't turned false when the callback is invoked.

### Analysis 

This problem was discovered in an internal test. The test is very simple, something like
```dart
  testWidgets('Culprit test', (tester) async {
    runApp(LocaleApp());
    await tester.pump();
  });
```
where `LocaleApp` is a widget that uses the localization API provided by `MaterialApp`.

When the test is run **on a live device**, it crashes with the following log:
```
'package:flutter/src/widgets/binding.dart': Failed assertion: line 968 pos 16: 'sendFramesToEngine': is not true.
When the exception was thrown, this was the stack:
dart:core-patch/errors_patch.dart 51:61            _AssertionError._doThrowNew
dart:core-patch/errors_patch.dart 40:5             _AssertionError._throwNew
package:flutter/src/widgets/binding.dart 968:16    WidgetsBinding.drawFrame.<fn>
package:flutter/src/scheduler/binding.dart 343:19  SchedulerBinding._executeTimingsCallbacks
dart:ui/hooks.dart 328:13                          _invoke1
dart:ui/platform_dispatcher.dart 611:5             PlatformDispatcher._reportTimings
dart:ui/hooks.dart 278:31                          _reportTimings
```

Deeper analysis shows that this problem is a combination of `testWidgets`, live device testing, `runApp`, and `deferFirstFrame`.

The `sendFramesToEngine` can turn false only if `deferFirstFrame` is called when no frame has been rendered yet. But if no frame has been rendered yet, how can there be a timing reporting callback, which is only invoked when a frame is rendered?

This is because `testWidgets` does some preparation before running a test body: It pumps a dummy widget, then calls `resetFirstFrameSent`.

https://github.com/flutter/flutter/blob/7f2b238c55a191567f6e93a170419aed41d8eb62/packages/flutter_test/lib/src/binding.dart#L1006-L1013

The pumped dummy widget is guaranteed rendered, and leaves a timing reporting callback on hold. This creates an inconsistent state at start of the test body: there's a timing report callback pending (since the UI thread has not `await`ed yet), but the framework thinks that no frame is sent. 

Now, if the test body uses `runApp` (instead of `tester.pump`) so that the widget tree is attached but no frame is rendered, and builds the widget tree, which defers the first frame, and then await, then the timing report callback will be invoked, finding the current value of `sendFramesToEngine` is false!

Apparently, the `testWidgets` leaking the timing reporting callback from the preparation is also a problem, and fixing either `testWidgets` or `WidgetTester` can solve the crash, but the `WidgetTester` should be of higher priority because:
1. `WidgetTester` is a production class
2. The assumption of `WidgetTester` is 100% wrong
3. It might not be trivial to fix: we might not be able to simply waiting for `waitUntilFirstFrameRasterized` since a warm-up frame might never be rendered.

This issue might be related to https://github.com/flutter/flutter/issues/132969
